### PR TITLE
initial cache driver support

### PR DESCRIPTION
**Pull request by avoltz**
Initial support of the cache volumes

To install the driver, first build the docker image:
```
make && docker build -t blob-csi -f pkg/blobplugin/Dockerfile
```
Then push the image to a container registry and install from the custom image:
```
helm install --set-string image.blob.repository=<registry>/blob-csi --set-string image.blob.tag=latest -n kube-system blob-csi-driver charts/latest/blob-csi-driver/
```

To configure create a storageclass like:
```
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: blob-cache
provisioner: blob.csi.azure.com
parameters:
  iscacheenabled: "true"
volumeBindingMode: Immediate
```

and pod + volume claim:
```
---
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: pvc-blob
spec:
  accessModes:
    - ReadWriteMany
  resources:
    requests:
      storage: 10Gi
  storageClassName: blob-cache
---
kind: Pod
apiVersion: v1
metadata:
  name: busybox
spec:
  containers:
    - image: busybox
      name: bb
      command:
        - sleep
        - infinity
      volumeMounts:
        - name: blob
          mountPath: "/mnt/blob"
  volumes:
    - name: blob
      persistentVolumeClaim:
        claimName: pvc-blob
```

You can follow logs on the node plugin to observe the volume handling:
```
kubectl logs -f -n kube-system -l app=csi-blob-node -c blob
```

### CODE REVIEW

File: charts/latest/blob-csi-driver/templates/csi-blob-node.yaml
Comment (thhicks) [inline charts/latest/blob-csi-driver/templates/csi-blob-node.yaml]: parameter rather than hard coded?

File: deploy/csi-blob-node.yaml
Comment (thhicks) [inline deploy/csi-blob-node.yaml:119]: same on parameters rather than hard coded

File: charts/latest/blob-csi-driver/templates/csi-blob-node.yaml
Comment (thhicks) [inline charts/latest/blob-csi-driver/templates/csi-blob-node.yaml:36]: i believe this is needed because of the EDGE_CACHE_CONFIG_ENDPOINT.  If you parameterize that, is there a way for the appropriate service fqdn to be filled in by helm rather than changing the dnsPolicy?

File: pkg/blob/nodeserver.go
Comment (thhicks) [inline pkg/blob/nodeserver.go:263]: since the existing blob fuse impl needs the key, are the key being exchanged with blobfuse in a different manner?

File: pkg/blobplugin/main.go
Comment (thhicks) [inline pkg/blobplugin/main.go]: can this reference a different service (ex. hydra-config)?

File: pkg/edgecache/manager.go
Comment (thhicks) [inline pkg/edgecache/manager.go:48]: cosmetic: is there significant to the GetStagingPath vs sendGetVolume from a first-letter-capitalized perspective?

Comment (avoltz) [reply pkg/edgecache/manager.go:48]: In go, 'public' functions in a package are capitalized. Internal functions are not, so the interface functions in this package use capital first letters.

File: charts/latest/blob-csi-driver/templates/csi-blob-node.yaml
Comment (avoltz) [reply charts/latest/blob-csi-driver/templates/csi-blob-node.yaml:36]: We could use params, but we will need this DNS policy regardless. Without it, the service is not resolvable even by fqdn.

File: deploy/csi-blob-node.yaml
Comment (avoltz) [reply deploy/csi-blob-node.yaml:119]: There do not appear to be any params in the deploy tree. I'm not sure how this section of the tree is used currently, the addon pr appears to fork this yaml content.

We could update `install-driver.sh` to use sed. There is some prior art there, but I don't know if we should really.

We can do a param in helm though.

File: charts/latest/blob-csi-driver/templates/csi-blob-node.yaml
Comment (thhicks) [reply charts/latest/blob-csi-driver/templates/csi-blob-node.yaml:36]: This is required by grpc communications internal to the edgecache implementation  then?

Comment (avoltz) [reply charts/latest/blob-csi-driver/templates/csi-blob-node.yaml:36]: This is required so that csi-blob-node can send `CreateBlob/GetBlob`.

The other communications to initiate mount with the cached pod is done thru a unix socket under /csi/nfsd_mounts.sock

File: pkg/blob/nodeserver.go
Comment (avoltz) [reply pkg/blob/nodeserver.go:263]: GetAuthEnv returns the `authEnv` value  which contains environ variables passed to the blobfuse process. The key is in those env values. This code is around line 470 in pkg/blob/blob.go

File: charts/latest/blob-csi-driver/templates/csi-blob-node.yaml
Comment (avoltz) [reply charts/latest/blob-csi-driver/templates/csi-blob-node.yaml]: fixed in latest update https://github.com/avoltz/blob-csi-driver/compare/e1cd53fd0b9faee30e6edcb872c2ff60dd077eca..0962931263e4f4df797f374d9db63f12bbf41edc

File: pkg/blobplugin/main.go
Comment (avoltz) [reply pkg/blobplugin/main.go]: fixed in latest update https://github.com/avoltz/blob-csi-driver/compare/e1cd53fd0b9faee30e6edcb872c2ff60dd077eca..0962931263e4f4df797f374d9db63f12bbf41edc

**Comment by thhicks** [comment]
LGTM

File: pkg/edgecache/manager.go
Comment (thhicks) [inline pkg/edgecache/manager.go:31]: cosmetic:  some commentary as to why this suffixing is necessary may be useful for future maintainers

**Review by thhicks — DISMISSED** [review]
LGTM
cosmetic comment on describing why suffixing is necessary

File: pkg/edgecache/manager.go
Comment (avoltz) [reply pkg/edgecache/manager.go:31]: added comment in latest update, looks like it reset your approval.

File: pkg/blob/nodeserver.go
Comment (andyzhangx) [inline pkg/blob/nodeserver.go:306]: this is also creating a volume in NodeStageVolume, it's not appropiate to create volume here since driver running on the agent node may probably do not have permission to create volume.

File: charts/latest/blob-csi-driver/templates/csi-blob-node.yaml
Comment (andyzhangx) [reply charts/latest/blob-csi-driver/templates/csi-blob-node.yaml:36]: do you have more details of following config:
```
            - name: EDGE_CACHE_CONFIG_ENDPOINT
              value: hydrad-config:8675
```

We hit some issues in the before when set as `ClusterFirstWithHostNet` in AKS managed control plane, that's why we changed to `Default` in the end.

File: pkg/blob/nodeserver.go
Comment (avoltz) [reply pkg/blob/nodeserver.go:306]: To clarify, this is creating an edge cache volume, not a kubernetes volume resource. We have to request the edge cache create a mount point for this blob storage resource.

I believe we need this in NodeStageVolume for static provisioning cases. When a statically provisioned PV is requested by the kube op, the first time the CSI driver will be notified is during nodestage volume. When caching is requested, we need to notify the edge cache services to create the mount point, then follow the alternate GRPC-based mount path.

I'm going to send along our design doc as well which should clarify some of this.

File: charts/latest/blob-csi-driver/templates/csi-blob-node.yaml
Comment (avoltz) [reply charts/latest/blob-csi-driver/templates/csi-blob-node.yaml:36]: When `iscacheenabled` is set for a volume, the node will mount the blob storage through the edge cache. In order for edge cache to provide a mount point, we must send a GRPC request to edge cache config service to configure the endpoint.

`hydrad-config` is the name (for now) of the config service for edge cache which is a clusterip service in kube.

When I left dnsPolicy: Default, I couldn't resolve this kube service, even by fqdn.

What issues did you see with ClusterFirst enabled?

File: pkg/edgecache/manager.go
Comment (brsmyth) [inline pkg/edgecache/manager.go:135]: Is the `mntctx.Err` case handled correctly at this point?

**Review by brsmyth — DISMISSED** [review]
I think this is ready to iterate on. 👍

File: pkg/edgecache/manager.go
Comment (avoltz) [reply pkg/edgecache/manager.go:135]: I think so, at this point mntctx is used to make the enclosed goroutine (line 117) cancellable. 

When time.After(timeout) is hit, we return from the function which triggers the deferred cancellation (line 111). that sets an error on the context allowing the unnamed func to exit.

if we comment out the unit test log silencing code the addmount cancel behavior is shown in logging during the test where cancel is met.

```
=== RUN   TestSendMount/CancelsEventually
W0524 12:10:00.603986   25621 manager.go:121] AddMount GRPC failed (will retry) returned with an error: rpc error: code = DeadlineExceeded desc =
W0524 12:10:00.604023   25621 manager.go:121] AddMount GRPC failed (will retry) returned with an error: rpc error: code = DeadlineExceeded desc =
...
W0524 12:10:00.611406   25621 manager.go:121] AddMount GRPC failed (will retry) returned with an error: rpc error: code = DeadlineExceeded desc =
W0524 12:10:00.611414   25621 manager.go:121] AddMount GRPC failed (will retry) returned with an error: rpc error: code = DeadlineExceeded desc =
--- PASS: TestSendMount (0.01s)
    --- PASS: TestSendMount/WorksFirstTry (0.00s)
    --- PASS: TestSendMount/WorksWithRetries (0.00s)
    --- PASS: TestSendMount/CancelsEventually (0.01s)
W0524 12:10:00.611423   25621 manager.go:121] AddMount GRPC failed (will retry) returned with an error: rpc error: code = DeadlineExceeded desc =
E0524 12:10:00.639681   25621 manager.go:128] AddMount cancelled "context canceled"
```

prior to getting this cancellation right i saw different errors when the client connection was closed. the func did not stop and would throw errors basically a leaked goroutine.

does that clarify?

Comment (brsmyth) [reply pkg/edgecache/manager.go:135]: Thanks! I will stare at it further to help my newbie understanding of the goroutine.

File: pkg/blob/nodeserver.go
Comment (avoltz) [reply pkg/blob/nodeserver.go:306]: Is it okay to resolve this? Hoping to get this PR merged and iterate further on our staging branch.

File: pkg/edgecache/manager.go
Comment (avoltz) [reply pkg/edgecache/manager.go:31]: ok to resolve?

File: pkg/blob/controllerserver.go
Comment (andyzhangx) [inline pkg/blob/controllerserver.go]: I think it's better add a new mount type, we already have `fuse`, `nfs`, this is the third mount type: `edgecache`
https://github.com/kubernetes-sigs/blob-csi-driver/blob/02ba84f85e6ac1ff153a7648297a91425e1db7ed/pkg/blob/blob.go#L118

File: pkg/blob/controllerserver.go
Comment (andyzhangx) [inline pkg/blob/controllerserver.go]: Are current account types all supported? `Standard_LRS, Premium_LRS, Standard_GRS, Standard_RAGRS`

**Review by andyzhangx — COMMENTED** [review]
is it ok to send a PR in the upstream repo? https://github.com/kubernetes-sigs/blob-csi-driver
And with the upstream repo PR, there are quite a lot tests to make sure there is no regression with this PR.

File: pkg/blob/nodeserver.go
Comment (andyzhangx) [reply pkg/blob/nodeserver.go:306]: I am ok with this, while if you merge the PR to a staging branch, are you going to make upstream changes finally? Maybe it's better merge this to upstream first if edge cache service is already ready?

File: pkg/blob/controllerserver.go
Comment (thhicks) [reply pkg/blob/controllerserver.go]: +1 on this suggestion, and +10 if it eliminates the need for the special suffixes on edge cache mount paths.

**Review by thhicks — COMMENTED** [review]
previous issues lgtm, i'm interested in seeing if there is a simplification by introducing a new mount type.   waiting for feedback on that before approving.

File: pkg/blob/controllerserver.go
Comment (avoltz) [reply pkg/blob/controllerserver.go]: The latest update removes this param and uses a new protocol option "edgecache"

See change here: https://github.com/avoltz/blob-csi-driver/pull/2/commits/e7c21ba6d2aceabbbfefdf09f6bb38f896c7cb00

File: pkg/blob/controllerserver.go
Comment (avoltz) [reply pkg/blob/controllerserver.go]: This is TBD from the edge cache side, so this is a callout for that future work.

File: pkg/blob/nodeserver.go
Comment (avoltz) [reply pkg/blob/nodeserver.go:306]: Currently edge cache is still in heavy development so we aren't ready to upstream yet.

Comment (andyzhangx) [reply pkg/blob/nodeserver.go:306]: sure, let me know when you are ready.

**Review by thhicks — APPROVED** [review]

**Review by brsmyth — APPROVED** [review]

**Pull request merged**
